enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -0,0 +1,310 @@
+
+Future types may be:
+* `events.kubernetes` - Kubernetes events
+* `audit.container-engine` - Container engine audit events
+
+There are no assumptions regarding whether or not an endpoint is deployed on or off cluster.  Endpoints off-cluster may require adminstrators to perform additional actions in order for logs to be forwarded (e.g. secret creation, opening port) 
+Following is the list of supported endpoint types for this proposal:
+
+* Elasticsearch
+* Fluent secure-forward
+
+### User Stories
+
+#### As an OKD admin, I want to collect logs and send them to my own aggregation solution, for example Splunk
+
+This is a typical example of organizations that desires to re-use their existing enterprise log solution.  We will succeeded if we are able to write logs to their logging service.
+
+#### As an OKD admin, I want to aggregate application logs on-cluster and infra logs off-cluster
+
+This is an example of an OKD cluster hosting solution where several organizations are each provided with a dedicated cluster.  The organization requires access to application container logs but the host requires access to the infra structure logs.
+
+### Implementation Details
+
+#### Assumptions
+
+* A pipeline shall support at least one target endpoint. 
+* A pipeline may support multiple target endpoints.
+* Endpoints that require secrets which are not created by the cluster-logging-operator shall be created and managed by the administrator of the endpoint
+* Secrets shall have keys of: cert,key,cacert which point to the respective certificates for which they represent
+
+#### Security
+* A pipeline may support TLS configuration by declaring a secret. Secrets must contain keys: cert, key, cacert
+* Endpoints defined without a secret will use unsecure communication
+
+#### Cluster Logging With Logstore, Collector, and LogForwarding
+* Undeclared source messages are routed to logstore
+* Declared source messages are routed to the defined service
+#### Cluster Logging With Pipelines Only
+* Undeclared source messages are ignored
+* Declared source messages are routed to the defined service
+
+
+### Risks and Mitigations
+
+## Design Details
+
+The `cluster-logging-operator` will use the forwarding API to generate a collector configuration and mount any secrets into the collector pod.
+
+### API Specification
+The proposed forwarding API:
+
+```
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "ClusterLogging"
+spec:
+  forwarding:
+    targets:
+     - type: “elasticsearch”
+       name: elasticsearch
+       endpoint: elasticsearch.svc.messaging.cluster.local
+       certificates:
+            secretName: elasticsearch
+     - type: “secureforward”
+       name: secureforward-offcluster
+       endpoint: https://secureforward.offcluster.com:9200
+       certificates:
+            secretName: secureforward
+    sources:
```

Comment:
or perhaps this should be called "pipelines"?

(where a pipeline consists of a source, and one or more targets.  or better, one or more sources and one or more targets)

Reply:
i think for a pipeline there are sources and targets ( * -> * ) where the source is the collector source (app logs, infra node logs, infra container logs, remote sources, etc) and the target is where we ship them to by specifying the target on the source definition.